src/nvim/highlight_group.c

```diff
@@ -1521,6 +1522,228 @@ static void highlight_list_one(const int id)
+      } else {
+        PUT_C(hl, "sp", INTEGER_OBJ(ae.rgb_sp_color));
+      }
+    }
+
+    if (ae.hl_blend > -1) {
+      PUT_C(hl, "blend", INTEGER_OBJ(ae.hl_blend));
+    }
+
+    if (mask & HL_BOLD) {
+      PUT_C(hl, "bold", BOOLEAN_OBJ(true));
+    }
+
+    if (mask & HL_STANDOUT) {
+      PUT_C(hl, "standout", BOOLEAN_OBJ(true));
+    }
+
+    switch (mask & HL_UNDERLINE_MASK) {
+    case HL_UNDERLINE:
+      PUT_C(hl, "underline", BOOLEAN_OBJ(true));
+      break;
+
+    case HL_UNDERCURL:
+      PUT_C(hl, "undercurl", BOOLEAN_OBJ(true));
+      break;
+
+    case HL_UNDERDOUBLE:
+      PUT_C(hl, "underdouble", BOOLEAN_OBJ(true));
+      break;
+
+    case HL_UNDERDOTTED:
+      PUT_C(hl, "underdotted", BOOLEAN_OBJ(true));
+      break;
+
+    case HL_UNDERDASHED:
+      PUT_C(hl, "underdashed", BOOLEAN_OBJ(true));
+      break;
+    }
+
+    if (mask & HL_STRIKETHROUGH) {
+      PUT_C(hl, "strikethrough", BOOLEAN_OBJ(true));
+    }
+
+    if (mask & HL_ITALIC) {
+      PUT_C(hl, "italic", BOOLEAN_OBJ(true));
+    }
+
+    if (mask & HL_INVERSE) {
+      PUT_C(hl, "reverse", BOOLEAN_OBJ(true));
+    }
+
+    if (mask & HL_NOCOMBINE) {
+      PUT_C(hl, "nocombine", BOOLEAN_OBJ(true));
+    }
+
+    mask = ae.cterm_ae_attr;
+
+    if (ae.cterm_fg_color != 0) {
+      PUT_C(hl, "ctermfg", INTEGER_OBJ(ae.cterm_fg_color - 1));
+    }
+
+    if (ae.cterm_bg_color != 0) {
+      PUT_C(hl, "ctermbg", INTEGER_OBJ(ae.cterm_bg_color - 1));
+    }
+
+    if (!(mask & HL_UNDERLINE_MASK) && ae.cterm_ae_attr != 0) {
+      // Ordered by `:h highlight-cterm`
+      Dictionary cterm = arena_dict(arena, HLATTRS_DICT_SIZE);
+      mask = ae.cterm_ae_attr;
+      if (mask & HL_BOLD) {
+        PUT_C(cterm, "bold", BOOLEAN_OBJ(true));
+      }
+
+      switch (mask & HL_UNDERLINE_MASK) {
+      case HL_UNDERLINE:
+        PUT_C(cterm, "underline", BOOLEAN_OBJ(true));
+        break;
+
+      case HL_UNDERCURL:
+        PUT_C(cterm, "undercurl", BOOLEAN_OBJ(true));
+        break;
+
+      case HL_UNDERDOUBLE:
+        PUT_C(cterm, "underdouble", BOOLEAN_OBJ(true));
+        break;
+
+      case HL_UNDERDOTTED:
+        PUT_C(cterm, "underdotted", BOOLEAN_OBJ(true));
+        break;
+
+      case HL_UNDERDASHED:
+        PUT_C(cterm, "underdashed", BOOLEAN_OBJ(true));
+        break;
+      }
+
+      if (mask & HL_STRIKETHROUGH) {
+        PUT_C(cterm, "strikethrough", BOOLEAN_OBJ(true));
+      }
+
+      if (mask & HL_INVERSE) {
+        PUT_C(cterm, "reverse", BOOLEAN_OBJ(true));  // inverse = reverse
+      }
+
+      if (mask & HL_ITALIC) {
+        PUT_C(cterm, "italic", BOOLEAN_OBJ(true));
+      }
+
+      if (mask & HL_STANDOUT) {
+        PUT_C(cterm, "standout", BOOLEAN_OBJ(true));
+      }
+
+      if (mask & HL_ALTFONT) {
+        PUT_C(cterm, "altfont", BOOLEAN_OBJ(true));
+      }
+
+      if (mask & HL_NOCOMBINE) {
+        PUT_C(cterm, "nocombine", BOOLEAN_OBJ(true));
+      }
```

Comment:
It seems that there is a lot of code duplicate between this branch and `hlattrs2dict()` in highlight.c. I guess it's better to add a `use_hex` argument to `hlattrs2dict()` instead.

Reply:
`use_hex` might not be the best name for this since RGB representation depends on `coloridx_to_name` function and `sg_rgb_fg_idx`, `sg_rgb_bg_idx`, `sg_rgb_sp_idx` which can only be found in `src/nvim/highlight_group.c` while `hlattrs2dict` was in `src/nvim/highlight_group.c`